comm/TLS.cc

```diff
@@ -0,0 +1,258 @@
+/**
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017 Intel Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction,
+ * including without limitation the rights to use, copy, modify,
+ * merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+#include <memory>
+
+#include <openssl/err.h>
+
+#include "ExceptionComm.h"
+#include "TLS.h"
+
+std::string bio_to_string(const std::unique_ptr<BIO, decltype(&BIO_free)>& bio);
+std::unique_ptr<BIO, decltype(&BIO_free)> string_to_bio(const std::string& string);
+
+OpenSSLInitializer::OpenSSLInitializer()
+{
+    OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS | OPENSSL_INIT_LOAD_CRYPTO_STRINGS, nullptr);
+
+    // Disable compression to avoid CRIME attacks
+    sk_SSL_COMP_zero(SSL_COMP_get_compression_methods());
+}
+
+OpenSSLInitializer& OpenSSLInitializer::instance()
+{
+    static OpenSSLInitializer initializer{};
+
+    return initializer;
+}
+
+std::string bio_to_string(const std::unique_ptr<BIO, decltype(&BIO_free)>& bio)
```

Comment:
True but my motivation was to prevent us, developers, from playing with raw pointers.

What do you think of typedef-ing it so it's better to the eyes: `typedef std::unique_ptr<BIO, decltype(&BIO_free)> BIOUniquePtr;`?

Reply:
> What do you think of typedef-ing it so it's better to the eyes: `typedef std::unique_ptr<BIO, decltype(&BIO_free)> BIOUniquePtr;`?

I don't object to adding a `typedef`, though I think it's also fine to keep it as is. At first glance this method seemed over-specified in that all it really needs is a `BIO` object to read from, and it doesn't really care how that object is managed by the caller.  But it sounds like you're deliberately adopting the convention that `BIO` objects should always be wrapped this way for safety, which makes sense.  From that perspective it's better to build this interface to accept only properly-wrapped instances, as you have it already. 👍 

Just as an aside re: raw pointers: I think that there can be a legitimate place for raw pointers in c++11 semantically to indicate a nullable reference.  In other words, it would be reasonable to use a raw pointer to indicate an externally managed object that may not exist, so be sure to null-check it, but don't `free` it or treat it as a degenerate array.  But that only works if we agree to adopt such a convention and use it consistently, which we don't currently.  Otherwise, I agree that they're probably best avoided.

Reply:
make sense. I think we more or less adopt that practice on aperturedb, but we haven't been policing it enough. I will start paying more attention to it. In this case, it makes sense to leave it as is.